src/Microsoft.ML.Recommend/MatrixFactorizationPredictor.cs

```diff
@@ -0,0 +1,468 @@
+        {
+            Contracts.CheckValue(env, nameof(env));
+            env.CheckValue(ctx, nameof(ctx));
+            ctx.CheckAtModel(GetVersionInfo());
+            return new MatrixFactorizationPredictor(env, ctx);
+        }
+
+        public void Save(ModelSaveContext ctx)
+        {
+            ctx.CheckAtModel();
+            ctx.SetVersionInfo(GetVersionInfo());
+
+            // *** Binary format ***
+            // int: number of rows (m), the limit on y
+            // ulong: Minimum value of the y key-type
+            // int: number of columns (n), the limit on x
+            // ulong: Minimum value of the x key-type
+            // int: internal dimension of matrices (k)
+            // Single[m * k]: the row dimension factor matrix P
+            // Single[k * n]: the column dimension factor matrix Q
+
+            _host.Assert(_m > 0);
+            _host.Assert(_n > 0);
+            _host.Assert(_k > 0);
+            ctx.Writer.Write(_m);
+            ctx.Writer.Write((InputYType as KeyType).Min);
+            ctx.Writer.Write(_n);
+            ctx.Writer.Write((InputXType as KeyType).Min);
+            ctx.Writer.Write(_k);
+            _host.Assert(Utils.Size(_p) == _m * _k);
+            _host.Assert(Utils.Size(_q) == _n * _k);
+            Utils.WriteSinglesNoCount(ctx.Writer, _p, _m * _k);
+            Utils.WriteSinglesNoCount(ctx.Writer, _q, _n * _k);
+        }
+
+        public void SaveAsText(TextWriter writer, RoleMappedSchema schema)
+        {
+            writer.WriteLine("# Imputed matrix is P * Q'");
+            writer.WriteLine("# P in R^({0} x {1}), rows correpond to Y item", _m, _k);
+            for (int i = 0; i < _p.Length; ++i)
+            {
+                writer.Write(_p[i].ToString("G"));
+                if (i % _k == _k - 1)
+                    writer.WriteLine();
+                else
+                    writer.Write('\t');
+            }
+            writer.WriteLine("# Q in R^({0} x {1}), rows correpond to X item", _n, _k);
+            for (int i = 0; i < _q.Length; ++i)
+            {
+                writer.Write(_q[i].ToString("G"));
+                if (i % _k == _k - 1)
+                    writer.WriteLine();
+                else
+                    writer.Write('\t');
+            }
+        }
+
+        private ValueGetter<Float> GetGetter(ValueGetter<uint> xGetter, ValueGetter<uint> yGetter)
+        {
+            _host.AssertValue(xGetter);
+            _host.AssertValue(yGetter);
+
+            uint x = 0;
+            uint y = 0;
+
+            var mapper = GetMapper<uint, uint, Float>();
+            ValueGetter<Float> del =
+                (ref Float value) =>
+                {
+                    xGetter(ref x);
+                    yGetter(ref y);
+                    mapper(ref x, ref y, ref value);
+                };
+            return del;
+        }
+
+        public ValueMapper<TXIn, TYIn, TOut> GetMapper<TXIn, TYIn, TOut>()
```

Comment:
Please add xmldocs to all `public` methods and types.

Reply:
Added. Thanks.

---
In reply to: [226368643](https://github.com/dotnet/machinelearning/pull/1263#discussion_r226368643) [](ancestors = 226368643)